frontend/src/components/pages/RegistrantExperience/RegistrationSteps/ReviewRegistration/EditContactCard.tsx

```diff
@@ -0,0 +1,399 @@
+import React, { useState, useMemo } from "react";
+import {
+  Box,
+  Button,
+  Divider,
+  Flex,
+  FormControl,
+  FormErrorMessage,
+  FormLabel,
+  Heading,
+  Input,
+  Spacer,
+  Text,
+  Textarea,
+  useDisclosure,
+  useToast,
+  Wrap,
+  WrapItem,
+} from "@chakra-ui/react";
+import {
+  PersonalInfoActions,
+  PersonalInfoReducerDispatch,
+} from "../../../../../types/PersonalInfoTypes";
+import { EmergencyContact } from "../../../../../types/CamperTypes";
+import RequiredAsterisk from "../../../../common/RequiredAsterisk";
+import DeleteModal from "../../../../common/DeleteModal";
+import {
+  checkEmail,
+  checkFirstName,
+  checkLastName,
+  checkPhoneNumber,
+  checkRelationToCamper,
+} from "../PersonalInfo/personalInfoReducer";
+
+type EditContactCardProps = {
+  contact: EmergencyContact;
+  contactIndex: number;
+  dispatchPersonalInfoAction: (action: PersonalInfoReducerDispatch) => void;
+};
+
+const EditContactCard = ({
+  contact,
+  contactIndex,
+  dispatchPersonalInfoAction,
+}: EditContactCardProps): React.ReactElement => {
+  const [updateMemo, setUpdateMemo] = useState(0);
+  const initialContact: EmergencyContact = useMemo(() => contact, [updateMemo]);
+
+  const [editing, setEditing] = useState(false);
+
+  const [isFirstNameInvalid, setIsFirstNameInvalid] = useState<boolean>(false);
+  const [isLastNameInvalid, setIsLastNameInvalid] = useState<boolean>(false);
+  const [isEmailInvalid, setIsEmailInvalid] = useState<boolean>(false);
+  const [isPhoneNumberInvalid, setIsPhoneNumberInvalid] = useState<boolean>(
+    false,
+  );
+  const [isRelationInvalid, setIsRelationInvalid] = useState<boolean>(false);
+
+  const updateFormErrorMsgs = () => {
+    // Check if we're on secondary contact. If so, then we display error messages only if user has started filling secondary contact.
```

Comment:
I am assuming you got this from the `checkPersonalInfoFilled()` function in the `frontend/src/components/pages/RegistrantExperience/RegistrationSteps/PersonalInfo/personalInfoReducer.tsx` file? 

If you just copied and pasted, could we modify that function to use 2 helper functions: 
1. helper method to check for camper details:
```
 // Check camper card
    if (
      !(
        checkFirstName(camper.firstName) &&
        checkLastName(camper.lastName) &&
        checkAge(camper.age, camp.ageUpper, camp.ageLower)
      )
    )
      return false;
```
2. method for checking campers:
```
// Check contact cards
    if (camper.contacts.length > 2 || camper.contacts.length < 1) {
      return false; // Need to have either 1 or 2 contacts only
    }

    // Check primary contact card
    const primaryContact = camper.contacts[0];
    if (
      !(
        checkFirstName(primaryContact.firstName) &&
        checkLastName(primaryContact.lastName) &&
        checkEmail(primaryContact.email) &&
        checkPhoneNumber(primaryContact.phoneNumber) &&
        checkRelationToCamper(primaryContact.relationshipToCamper)
      )
    ) {
      return false;
    }

    // Check secondary contact card
    if (camper.contacts.length > 1) {
      const secondaryContact = camper.contacts[1];
      if (
        (secondaryContact.firstName ||
          secondaryContact.lastName ||
          secondaryContact.email ||
          secondaryContact.phoneNumber ||
          secondaryContact.relationshipToCamper) &&
        !(
          checkFirstName(secondaryContact.firstName) &&
          checkLastName(secondaryContact.lastName) &&
          checkEmail(secondaryContact.email) &&
          checkPhoneNumber(secondaryContact.phoneNumber) &&
          checkRelationToCamper(secondaryContact.relationshipToCamper)
        )
      ) {
        return false;
      }
    }
```

Then we can just call those helper methods in these Edit cards instead of copying and pasting the logic

Reply:
For this, I did create some exported functions in the `personalInfoReducer.tsx` doing pretty much what you outlined, and committed these changes since I still think it's an improvement, at least for the personal info reducer. However I don't think there's a clear way to implement that in my two edit components- my edit logic also has to simultaneously set invalid fields to invalid, so we need to check fields one by one and can't do them all at once by using the functions you wrote above. Thoughts?

Reply:
Adding onto above, I noticed we can take your approach with the additional info reducer since the reducer is what takes care of setting fields to invalid, versus in this personal info reducer, we have to do it ourselves in the component we define it in. IMO it's probably too much work to change how the personal info reducer works now, so I rather just leave this as is